flatironinstitute/mountainlab-js
# refactor npm packaging to make less monolithic, support global installs (fixes #23)

This patch touches a lot of files, but it doesn't change any .js code. It only refactors the way things are built to be more 'npm-ish' (as I understand it, as a newbie). In particular:

1. **Local dependencies (mlproc, lari, mlstudy, kbucket_upload) are specified as such in the main package.json file** (rather than running `cd lari && npm install`... in a postinstall script), resolving #23 . This allows for: automatically installing binaries provided by the deps; deduplication across subdependencies; and ability to split these off as their own npm packages at some point. I also add these packages to the 'bundledDependencies' list so that `npm pack` or `npm publish` will bundle them up and ship them in one package correctly.
2. **Use npm's facilities for specifying binaries** so that: they can be installed automatically by "npm install --global"; and the binaries provided by dependencies are available to the main package at a well-specified location (./node_packages/.bin), even on local installs. This means we don't need to hardcode paths to specific binaries (which would otherwise break the model of the dependencies being real packages)
3. **Refactor the way the binary scripts (ml-config, ml-*) call the binaries**, to work in this new world. In particular, we need to  handle when the scripts are called from a symlink (since this is how the npm installer installs binaries), which means using `realpath $0` rather than `$0` (which resolves to the location of the symlink). Also, we now need to avoid using path traversal within the repo to call binaries (`../mlproc/mlproc`), which is brittle, and would break if we were to install dependencies from another location. Instead, we look for them at the location where they are guaranteed to be after the packages that provide them have been installed (`../node_packages/.bin/mlproc`). I chose to also provide `mlproc` `ml-run-study` and `lariserver` themselves as scripts in the main mountainlab-js package, since I think end-users may want to call these directly (rather than only ever using the helper scripts like `ml-config` (which just calls `mlproc config`). The helper scripts now just refer to these local helpers (`./mlproc`).
4. **Get ready to ship a true, slimmed-down "Production" build** (without devDependencies): Move 'commander' from 'devDependencies' to regular 'dependencies' in mlproc since it is required for a regular install; and no longer call webpack on mlstudy during a regular install. **_<==This may be an error on my part_**; I don't know why mlstudy is webpacked, or if this is depended on by any external code, but I couldn't find any reference to ./mlstudy/web/* in this repo, and removing it from a production install makes the installed data much much smaller. Note that webpack can still be called explicitly with `npm webpack mlstudy` (I also fixed the issue that caused this to fail if called from outside of the mlstudy directory itself.)
5. A couple minor cleanup items: ignore ./mlstudy/package-lock.json, add an .npmignore file to help me avoid packaging up a tarball in my tarball.

### How to test
For testing, I would suggest: removing all the mountainlab-js scripts/binaries from your path; do a fresh clone of this branch and trying to run `npm install -g .`. After it installs (assuming `npm prefix --global` is a directory on your path), you should be able to run all the provided scripts/binaries correctly. I have tested this on my Ubuntu machine. 

### Rationale/Future work
* This was all in service of being able to build a conda package (#14, #15), which should now be a vanilla affair. 
* Rather than providing scripts for `mlproc`, `mlstudy`, and `lariserver`, it would make more sense to do global installs of these subpackages. Then we could just call these from scripts like `ml-config`  without having to do any other path munging. But I have yet to figure out how to do this, and it doesn't much matter, functionally.
* npm recommends committing our package-lock.json files, but I don't quite understand this issue. For now harmonize on not using them.
